### DISCUSSION TITLE
Design Doc: AST Helper

### DISCUSSION THREAD

**Discussion by chriswr95** [graphql-java/graphql-java#2627] [Ideas]
## **Proposal**

Create an `AstHelper` class with well defined and tested methods to perform common operations on the abstract syntax tree. 

### **Methods**

- `getRecursivelyReferencedFragments()` - Given a node in the AST, return all fragments referenced by the node.
- `detectDeprecatedFields()` - Given a query and a reference to a graphql schema, detect if the query references any fields marked deprecated by the schema.

### **Implementation**

`getRecursivelyReferencedFragments()`

We don't want to traverse the whole tree here so we will use a modified depth first search. In pseudo code:

```jsx
List<Fragment> getRecursivelyReferencedFragments(Node operation) {
  List fragments = [];
  Set<String> collectedNames = Set<String>();
  Array<SelectionSetNode> nodesToVisit = Stack<Node>;
  nodesToVisit.push(operation);
  while ((node = nodesToVisit.pop())) {
		List<Node> fragmentSpreads = getFragmentSpreads(node);
    for (spread in fragmentSpreads) {
      const fragName = spread.name;
      if (collectedNames[fragName] !== true) {
        collectedNames[fragName] = true;
        fragments.push(spread);
        nodesToVisit.push(fragment.getChildren());
      }
    }
  }
  return fragments;
}

List<Fragment> getFragmentSpreads(Node root) {
	// Use NodeTraverser to walk the subtree, and return all fragment spreads
}
```

`detectDeprecatedFields()`

Detecting deprecated fields is more straightforwards. We can use the QueryTraverser visitor pattern. In (mostly) pseudo code:

```graphql
List<Node> detectDeprecatedFields(Document document, Schema graphqlSchema) {
	QueryTraverser queryTraverser = QueryTraverser.newQueryTraverser()
	                .document(document)
	                .operationName(operation)
	                .schema(graphQLSchema)
	                .variables(variables)
	                .build();
	
	final List<String> deprecatedFieldList = new ArrayList<>();
	queryTraverser.visitPreOrder(new QueryVisitorStub() {
	    @Override
	    public TraversalControl visitFieldWithControl(QueryVisitorFieldEnvironment fieldEnvironment) {
	        collectDeprecated(fieldEnvironment, deprecatedFieldList);
	        return TraversalControl.CONTINUE;
	    }
	});
	return deprecatedFieldList;
}
```

### Testing
We will write unit tests in Groovy to achieve 100% code coverage `AstHelper` class. I am unaware of any other testing requirements, please let me know if I am missing anything.

### **Extension**

We can evolve this class in the future to add more functionality. For example:

- `normalizeExecutableQuery()` - Given a query, return a normalized version. his represents the AST collapsed into its eventual "field" with schema types and so on all resolved.

**Open Questions**

Where in the directory structure should this class live?
My initial thought is `src/main/java/graphql/AstHelper.java`

**Comment by bbakerman** [comment]
> Where in the directory structure should this class live?
My initial thought is src/main/java/graphql/AstHelper.java

Most of the AST code is in `graphql.language`

> getRecursivelyReferencedFragments

Can you tell me more about why you want to know if fragments reference fragments?  Whats the use case for knowing this and what would you do if you had that list?

**Reply by chriswr95** [reply]
Yeah, a common pattern at Twitter is to store our queries and fragments in separate directories. Queries then reference some, but generally not all, the fragments. During build time we have a tool which parses the query, and attaches only the referenced fragments. My thought is that other devs may have similar workflows, or find other uses for knowing what fragments are referenced by a query.

For example on input:
```
fragment B on Query { b }
fragment A on Query { a }
{ ...A }
```
we expect to be able to drop the unused fragment so that the output looks like this:
```
fragment A on Query { a }
{ ...A }
```
